Migrate to Dart 3

bernaferrari opened the pull request:
Fix https://github.com/google/file.dart/issues/209

I also made sure `dart analyzer` result is perfect.

`abstract class` => `mixin`. They can't be `mixin class` because then they need to expose/re-implement dozens of methods.

![image](https://user-images.githubusercontent.com/351125/226994562-5d1f8653-a358-464b-b201-4edafdc47525.png)

bernaferrari committed:
Migrate to Dart 3.

Also apply Dart fixes.

gspencergoog reviewed (approved):
![32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d](https://user-images.githubusercontent.com/8867023/37737237-66e9c9dc-2d10-11e8-8797-13a405eace49.png)

gspencergoog commented:
> Test is failing because it is testing against Dart stable. Not sure we should change the test file to work against master, or just ignore it.

If we can change the file to work for both, that would be ideal.

gspencergoog commented:
> Greg will get tired of me appearing all day into his notifications. I hope @jakemac53 can review this.

LOL! No, I don't mind at all, especially when you're doing cleanups like this.

bernaferrari committed:
Update pubspec.

bernaferrari commented:
I guess you need to approve the workflow change for it to be executed.

bernaferrari commented on .github/workflows/file.yml:
```diff
@@ -17,12 +17,13 @@ jobs:
     steps:
-      - uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c
-      - uses: dart-lang/setup-dart@a57a6c04cf7d4840e88432aad6281d1e125f0d46
+      - uses: actions/@24cb9080177205b6e8c946b17badbe402adc938f
```
Yes, sharp eyes!

goderbauer commented:
Nice! @bernaferrari Any chance you want to send us a PR for https://github.com/flutter/flutter/issues/123172 as well once this is released? 😆

bernaferrari commented:
There it goes. Do I need to ask Hixie for exception or you can press a button to ignore that?

devoncarew commented:
We will want to rev. the pubspec version and add an associated changelog entry. The package is currently at `6.1.4`; @jakemac53 or @natebosch may have advice about whether this is a patch / minor / major version rev.

bernaferrari commented:
If we follow Dart semantic versioning, it should probably be `7.0.0` because it might be breaking for some people. But it is a lot less breaking than, say, null safety was. So could be 6.2 as well.

I updated and fixed the action. There is no need to fix anything, the CI is already working the way we want.

natebosch commented:
> because it might be breaking for some people

Breaking for which people in what ways?

bernaferrari commented:
TIL we might want an `abstract mixin class` because that allows extends (same behavior as right now).

I was using `mixin`, but that doesn't allow extend. Should we allow extends?

<img width="685" alt="image" src="https://user-images.githubusercontent.com/351125/227017222-0a2d4a44-8f7d-4b36-8dd6-3d76558bf4c6.png">

bernaferrari committed:
Use `abstract mixin class` instead of `mixin`.

jakemac53 commented:
Thanks for taking the time to fix this up! This PR does bring up a bunch of existential questions:

- Should these be mixin declarations or mixin classes? I think probably a bunch of them are only intended to be mixed in, and should be mixin declarations.
- In general, what do we want package authors to do if they do convert some `abstract class` delarations to `mixin` declarations? We have gone to pretty extreme lengths to _not_ have major ecosystem churn for this release (even rewriting SDK constraints). To end up having a ton of ecosystem churn just to convert a bunch of `abstract class` declarations to `mixin` declarations is not imo a good tradeoff. So we should either tell people to make them `mixin class` declarations or tell them to convert them to `mixin` declarations and not do a breaking change.
- At what point do we want to start publishing packages which rely on Dart 3.0, and what should the min SDK constraint for them be? We should at least ensure its a dev release that contains the flag flip.

jakemac53 commented on packages/file/pubspec.yaml:
```diff
@@ -1,19 +1,19 @@
 
 environment:
-  sdk: '>=2.12.0 <3.0.0'
+  sdk: '>=3.0.0-0 <4.0.0'
```
Fwiw this constraint will need to be more specific, something like `>=3.0.0-321.0.dev <4.0.0`. However, I am not totally sure how that would interact with the _beta_ channels, I don't know what their versions look like.

bernaferrari replied:
I think 3.0.0-0 works, but won't work with, say 319 lol... Does it need to be "that" gradual?

bernaferrari replied:
Flutter is using 3.0.0-0

mit-mit commented:
> At what point do we want to start publishing packages which rely on Dart 3.0, and what should the min SDK constraint for them be? We should at least ensure its a dev release that contains the flag flip.

I think we can start to do that as soon as there is a beta release that has the new features on (i.e. the upcoming beta release), and then we make that the min.

A little closer to stable -- but stil before stable -- we can publish the final stable versions that have `sdk: >=3.0.0`.

This should all work fine now that pub.dev supports "previews": https://dart.dev/tools/pub/publishing#publishing-previews

jakemac53 commented:
> This should all work fine now that pub.dev supports "previews": https://dart.dev/tools/pub/publishing#publishing-previews

This doesn't actually do anything other than show the version as a "preview" on pub right? So a bad constraint, or a breakage in the SDK prior to stable, could still cause all the typical problems?

devoncarew commented:
I think we could reasonable `3.0.0-0`. I _believe_ that will select all 3.0.0 dev releases, including ones that do and don't have the flag flipped, but that may not be an issue. There is probably an implicit assumption that people who are on dev should keep that channel up to date.

jakemac53 commented:
`>=3.0.0-0` will allow all pre-release versions so it will be broken for sure on some of them. I don't have a strong opinion on if that matters.

`>=3.0.0-321.0.dev` will only allow the dev releases that definitely work, but will still allow earlier betas which won't work, and won't allow edge releases afaik.

jakemac53 commented:
Actually, I just selected the beta channel in flutter and the version looks like `3.0.0-290.1.beta`, so the dev constraint given above should also work just fine, it seems the numbers are likely correlated and its just the suffix (`.beta` vs `.dev`) that changes. We could consider dropping the suffix too.

I don't care so much if edge releases don't work, they don't have meaningful numbers (its a sha and an `edge` prefix).

bernaferrari commented:
But the beta version in Flutter doesn't have support for class modifiers, they only flipped the switch a few days ago.. so only master has it.

jakemac53 commented:
> But the beta version in Flutter doesn't have support for class modifiers, they only flipped the switch a few days ago.. so only master has it.

Right, and that beta version won't be selectable by the dev constraint I suggested :). But the next beta will.

jakemac53 commented:
Flutter master is also on `3.0.0-356.0.dev` now so you _will_ be able to use it there.

bernaferrari committed:
Use `3.3.0-356.0.dev`.

jakemac53 commented on packages/file/pubspec.yaml:
```diff
@@ -7,10 +7,10 @@ description:
 
 environment:
-  sdk: '>=3.0.0-0 <4.0.0'
+  sdk: '>=3.0.0-356.0.dev <4.0.0'
```
Almost entirely academic, but this could be as low as `>=3.0.0-321.0.dev`. Realistically it isn't going to matter much so if you don't feel like updating it, don't :).

jakemac53 replied:
On second thought, `>=3.0.0-325.0.dev` is probably better because that will also enable patterns/records which technically could break some things (although we know they don't already, at least for this package).

bernaferrari replied:
For me, I would just let 3.0.0-0 and merge, but let's go.. lol

bernaferrari replied:
I have a random question, I'm not following Dart that closely, but can I already do `switch(a) => option1 => 1, option2 => 2`?

jakemac53 replied:
I think it looks a bit different than that but yes

bernaferrari committed:
Move to 321.

bernaferrari committed:
Move to 325.

bernaferrari commented:
The last thing for now is deciding if you want `abstract mixin class` (not breaking) or `mixin` (breaking, wouldn't be extendable anymore).

natebosch commented:
> * To end up having a ton of ecosystem churn just to convert a bunch of `abstract class` declarations to `mixin` declarations is not imo a good tradeoff. So we should either tell people to make them `mixin class` declarations or tell them to convert them to `mixin` declarations and not do a breaking change.

I think these are likely to be case by case.

Some package that isn't deep in most dependency trees without evidence of usage with `extends` should choose a major version bump and `mixin`. Some package that is very deep in the dependency tree with lots of existing usage of `extends` should choose a feature version bump an `mixin class`. In between it might not be clear.

The only choice that is off the table is a feature version bump with a switch to `mixin`.

For this particular case - the `*Mixin` names and the usage I see from a brief skim tend to push me towards `mixin`. I don't recall this package being particularly painful during the null safety major version bump. I don't have a strong preference for the breaking change though, and I'd be perfectly happy with a guidance to prefer non-breaking for this migration in general.

The safest path is no breaking change. Some time after Dart 3 we might be able to get analyzer support for deprecating the `class` part of a `mixin class` with an incremental migration and `dart fix` support.

bernaferrari commented:
On this specific code, I found zero `extends` usage in GitHub code search, and Flutter usages are all `with`.

jakemac53 commented:
> The only choice that is off the table is a feature version bump with a switch to `mixin`.

I really do not think that should be off the table. If there is no evidence of `extends`, as in this case, and the package isn't _super_ widely used, I think it should be generally acceptable to convert to a `mixin` declaration and not do a breaking change. That is in fact what I would personally do for this package.

jakemac53 committed:
only run tests on dev

bernaferrari committed:
Use mixin.

jakemac53 commented on packages/file_testing/pubspec.yaml:
```diff
@@ -7,7 +7,7 @@ environment:
 
 dependencies:
-  test: ^1.23.1
+  test: ^1.24.0
```
This probably can/should stay at ^1.23.1 - flutter pins test_api versions which means flutter users can't update `test` without updating their SDK.

bernaferrari committed:
Fix test.

jakemac53 commented:
Ok, sorry for all the delay here. We got in a room together to discuss our broader strategy for Dart 3 packages, and migrations to mixin _declarations_ from "mixinable" classes. In summary:

- We think we should go ahead and do a breaking change release in this package, and other packages, when migrating to mixin declarations.
- We want to hold off on merging/publishing this PR until after the final Beta is released (a couple weeks probably). At that point we will have much higher confidence that Dart 3 is more stable and won't have any breaking changes to the features within it (we are still actively making small tweaks right now).

Thanks for your hard work and patience @bernaferrari !

bernaferrari committed:
Make it breaking.

bernaferrari commented:
I made it breaking. We could put the other breaking change PR together in 7.0. Is the rollout automatic? I could add // NEXT.

jakemac53 commented on packages/file/pubspec.yaml:
```diff
@@ -1,13 +1,13 @@
 name: file
-version: 6.2.0
+version: 7.0.0
```
We can set the version to `7.0.0-wip` to indicate this isn't  yet ready to publish (until the other PR also lands). Same for the changelog. Then we could probably merge this any time.

bernaferrari committed:
Make it wip.

jakemac53 commented on .github/workflows/file.yml:
```diff
@@ -42,7 +42,7 @@ jobs:
         package: [file]
         os: [ubuntu-latest, macos-latest, windows-latest]
-        sdk: [stable, beta, dev]
+        sdk: [dev]
```
```suggestion
        sdk: [stable, dev]
```

bernaferrari committed:
Add stable.

jakemac53 commented:
@bernaferrari of course it never ends, we will need to add a dependency_override on glob I think to get a valid resolve :)

bernaferrari commented:
wouldn't test 1.25 solve? I don't understand exactly what is happening

jakemac53 commented:
> wouldn't test 1.25 solve? I don't understand exactly what is happening

The issue is the `glob` depends on `file` (version 6.x) https://github.com/dart-lang/glob/blob/master/pubspec.yaml#L12. And test depends on glob. There is no compatible version of glob with file 7.x, and so since this package depends on test (to run its tests), there is a version mismatch.

```yaml
dependency_overrides:
  glob: 2.1.1
```

Should solve it.

bernaferrari committed:
Fix test?

jakemac53 commented:
The main pubspec in `packages/file/pubspec.yaml` also needs an update 👍

bernaferrari commented:
oh crap lol, thanks

bernaferrari committed:
Fix forever.
